syntax = "proto3";

package inventory;

option go_package = "./inventorypb";

message Product {
  string id = 1;
  string name = 2;
  int32 quantity = 3;
}

message StockChange {
  string product_id = 1;
  int32 delta = 2;
  int64 timestamp = 3;
}

message AddProductRequest {
  string name = 1;
  int32 quantity = 2;
}

message AddProductResponse {
  Product product = 1;
}

message GetProductRequest {
  string id = 1;
}

message GetProductResponse {
  Product product = 1;
}

message UpdateStockRequest {
  string id = 1;
  int32 delta = 2;
}

message UpdateStockResponse {
  Product product = 1;
  StockChange change = 2;
}

message RemoveProductRequest {
  string id = 1;
}

message RemoveProductResponse {
  bool success = 1;
}

message ListProductsRequest {}

message ListProductsResponse {
  repeated Product products = 1;
}

message StreamStockAlertsRequest {
  int32 threshold = 1;
}

message StockAlert {
  Product product = 1;
  string message = 2;
}

service InventoryService {
  rpc AddProduct(AddProductRequest) returns (AddProductResponse);
  rpc GetProduct(GetProductRequest) returns (GetProductResponse);
  rpc UpdateStock(UpdateStockRequest) returns (UpdateStockResponse);
  rpc RemoveProduct(RemoveProductRequest) returns (RemoveProductResponse);
  rpc ListProducts(ListProductsRequest) returns (ListProductsResponse);
  rpc StreamStockAlerts(StreamStockAlertsRequest) returns (stream StockAlert);
}
